spec/ddtrace/logger_spec.rb

```diff
@@ -0,0 +1,263 @@
+        it { expect { Datadog::Tracer.log.warn('bar') { 'warn message' } }.to_not raise_error }
+      end
+
+      context 'default log level' do
+        before(:each) { log_warn() }
+        it { expect(lines.length).to eq(1) if lines.respond_to?(:length) }
+        it { expect(lines[0]).to match(/W, \[[0-9:T.-]+ #[0-9]+\]  WARN -- ddtrace: \[ddtrace\] warn message\n/) }
+      end
+
+      context 'debug log level' do
+        let(:log_level) { Logger::DEBUG }
+        before(:each) { log_warn() }
+
+        it { expect(lines.length).to eq(1) if lines.respond_to?(:length) }
+        # We add traceback information to the log line for errors
+        it do
+          expect(lines[0]).to match(Regexp.new('W, \[[0-9:T.-]+ #[0-9]+\]  WARN -- ddtrace: \[ddtrace\] '\
+                                               '\([\/a-z_]+\.rb:[0-9]+:in `log_warn\'\) warn message\n'))
+        end
+      end
+    end
+
+    context 'debug_logging enabled' do
+      after(:each) { Datadog::Tracer.debug_logging = false }
+
+      it do
+        Datadog::Tracer.debug_logging = true
+        expect(Datadog::Tracer.log.level).to eq(Logger::DEBUG)
+
+        Datadog::Tracer.debug_logging = false
+        expect(Datadog::Tracer.log.level).to eq(Logger::WARN)
+      end
+    end
+  end
+
+  describe 'custom logger' do
+    let(:logger) do
+      logger = Logger.new(buf)
+      logger.level = Logger::INFO
+
+      logger
+    end
+
+    # Invalid values to set
+    [
+      nil,
+      'this is a message'
+    ].each do |value|
+      context 'when it is #{value.inspect}' do
+        let(:logger) { value }
+
+        it { expect(Datadog::Tracer.log).to be(default_logger) }
+      end
+    end
+
+    context '#level' do
+      it { expect(Datadog::Tracer.log.level).to eq(Logger::INFO) }
+    end
+
+    context 'debug_logging enabled' do
+      after(:each) { Datadog::Tracer.debug_logging = false }
+
+      context 'when enabling' do
+        it do
+          Datadog::Tracer.debug_logging = true
+          expect(Datadog::Tracer.log.level).to eq(Logger::DEBUG)
+
+          Datadog::Tracer.debug_logging = false
+          expect(Datadog::Tracer.log.level).to eq(Logger::DEBUG)
+        end
+      end
+
+      context 'when disabling' do
+        it do
+          # Assert we are not enabled and have the default log level
+          expect(Datadog::Tracer.debug_logging).to be(false)
+          expect(Datadog::Tracer.log.level).to eq(Logger::INFO)
+
+          # Assert manually disabling it does not change log level
+          Datadog::Tracer.debug_logging = false
+          expect(Datadog::Tracer.log.level).to eq(Logger::INFO)
+        end
+      end
+    end
+  end
+
+  describe 'rate limiting' do
+    context 'when hitting rate limit' do
+      before(:each) do
+        # Stub Time.now to ensure it always returns the same value
+        #   so we stay in the same time bucket and everything gets grouped
+        allow(Time).to receive(:now).and_return Time.at(1550000000)
```

Comment:
This might be dangerous, stubbing a global interface for Time with a fixed value. Other mechanisms running in the test suite (including RSpec itself) might depend on calling this, and it could lead to unexpected behavior.

Might want to implement [Timecop or another time management gem](https://www.ruby-toolbox.com/categories/time_warping) to handle this.